airflow/sensors/smart_sensor_operator.py

```diff
@@ -0,0 +1,764 @@
+                    exception_info,
+                    is_infra_failure=is_infra_failure)
+            else:
+                self.cached_sensor_exceptions[cache_key] = SensorExceptionInfo(
+                    exception_info,
+                    is_infra_failure=is_infra_failure)
+
+            self._handle_poke_exception(sensor_work)
+
+    def flush_cached_sensor_poke_results(self):
+        """
+        Flush outdated cached sensor states saved in previous loop.
+
+        """
+        for key, cached_work in self.cached_dedup_works.items():
+            if cached_work.is_expired():
+                self.cached_dedup_works.pop(key, None)
+            else:
+                cached_work.state = None
+
+        for ti_key, sensor_exception in self.cached_sensor_exceptions.items():
+            if sensor_exception.fail_current_run or sensor_exception.is_expired():
+                self.cached_sensor_exceptions.pop(ti_key, None)
+
+    def poke(self, sensor_work):
+        """
+        Function that the sensors defined while deriving this class should
+        override.
+
+        """
+        cached_work = self.cached_dedup_works[sensor_work.cache_key]
+        if not cached_work.sensor_task:
+            init_args = dict(list(sensor_work.poke_context.items())
+                             + [('task_id', sensor_work.task_id)])
+            operator_class = import_string(sensor_work.op_classpath)
+            cached_work.sensor_task = operator_class(**init_args)
+
+        return cached_work.sensor_task.poke(sensor_work.poke_context)
+
+    def _emit_loop_stats(self):
+        try:
+            count_poke = 0
+            count_poke_success = 0
+            count_poke_exception = 0
+            count_exception_failures = 0
+            count_infra_failure = 0
+            for cached_work in self.cached_dedup_works.values():
+                if cached_work.state is None:
+                    continue
+                count_poke += 1
+                if cached_work.state == PokeState.LANDED:
+                    count_poke_success += 1
+                elif cached_work.state == PokeState.POKE_EXCEPTION:
+                    count_poke_exception += 1
+            for cached_exception in self.cached_sensor_exceptions.values():
+                if cached_exception.is_infra_failure and cached_exception.fail_current_run:
+                    count_infra_failure += 1
+                if cached_exception.fail_current_run:
+                    count_exception_failures += 1
+
+            Stats.gauge("smart_sensor_operator.poked_tasks", count_poke)
+            Stats.gauge("smart_sensor_operator.poked_success", count_poke_success)
+            Stats.gauge("smart_sensor_operator.poked_exception", count_poke_exception)
+            Stats.gauge("smart_sensor_operator.exception_failures", count_exception_failures)
+            Stats.gauge("smart_sensor_operator.infra_failures", count_infra_failure)
```

Comment:
Can you please add these metrics to https://github.com/apache/airflow/blob/master/docs/metrics.rst

Reply:
Thanks for reminding. Added metrics to the doc.